File: src/io/kyros/content/teleportation/TeleportTablets.java
package io.kyros.content.teleportation;

import io.kyros.Configuration;
import io.kyros.model.cycleevent.CycleEvent;
import io.kyros.model.cycleevent.CycleEventContainer;
import io.kyros.model.cycleevent.CycleEventHandler;
import io.kyros.model.entity.player.Boundary;
import io.kyros.model.entity.player.Player;

/**
 * 
 * @author Mack
 *
 */
public class TeleportTablets {

	public enum TabType {
		//		Standard
		HOME(8013, Configuration.RESPAWN_X, Configuration.RESPAWN_Y),
		VARROCK(8007, Configuration.VARROCK_X, Configuration.VARROCK_Y),
		LUMBRIDGE(8008, Configuration.LUMBY_X, Configuration.LUMBY_Y),
		FALADOR(8009, Configuration.FALADOR_X, Configuration.FALADOR_Y),
		CAMELOT(8010, Configuration.CAMELOT_X, Configuration.CAMELOT_Y),
		ARDOUGNE(8011, Configuration.ARDOUGNE_X, Configuration.ARDOUGNE_Y),


		//		Redirect Tablets
		RIMMINGTON(11741, 2956, 3217),
		TAVERLEY(11742, 2896, 3456),
		POLLNIVENEACH(11743, 3351, 2960),
		RELLEKKA(11744, 2643, 3676),
		BRIMHAVEN(11745, 2794, 3178),
		YANILLE(11746, 2606, 3098),
		TROLLHEIM(11747, Configuration.TROLLHEIM_X, Configuration.TROLLHEIM_Y),


		//		Arceuus Tablets
		ARCEUUS_LIBRARY(19613, 1646, 3806),
		DRAYNOR_MANNOR(19615, 3109, 3346),
		SALVE_GRAVEYARD(19619, 3434, 3461),
		FENKENSTRAIN_CASTLE(19621, 3548, 3529),
		WEST_ARDOUGNE(19623, 2538, 3305),
		HARMONY_ISLAND(19625, 3794, 2851),
		CEMETERY(19627, 2976, 3751),
		BARROWS(19629, 3565, 3306),
		APE_ATTOL(19631, 2757, 2781),


		//		Ancient Tablets
		PADDEWWA(12781, Configuration.PADDEWWA_X, Configuration.PADDEWWA_Y),
		SENNTISTEN(12782, Configuration.SENNTISTEN_X, Configuration.SENNTISTEN_Y),
		KHARYRLL(12779, Configuration.KHARYRLL_X, Configuration.KHARYRLL_Y),
		LASSAR(12780, Configuration.LASSAR_X, Configuration.LASSAR_Y),
		DAREEYAK(12777, Configuration.DAREEYAK_X, Configuration.DAREEYAK_Y),
		CARRALLANGER(12776, Configuration.CARRALLANGAR_X, Configuration.CARRALLANGAR_Y),
		ANNAKARL(12775, Configuration.ANNAKARL_X, Configuration.ANNAKARL_Y),
		GHORROCK(12778, Configuration.GHORROCK_X, Configuration.GHORROCK_Y),


//		Lunar
		MOONCLAN(24949, 2113,3915),
		OURANIA(24951, 3015, 5622),
		WATERBIRTH_ISLAND(24953,2527, 3740),
		BARBARIAN(24955, 2519, 3571),
		KHAZARD(24957, 2660, 3158),
		FISHING_GUILD(24959, 2635, 3425),
		CATHERBY(24961, 2804, 3433),
		ICE_PLATEU(24963, 2916, 3921),



//		Misc
		WILDY_RESOURCE(12409, 3184, 3945),
		PIRATE_HUT(12407, 3045, 3956),
		MAGE_BANK(12410, 2538, 4716),
		CALLISTO(12408, 3293, 3847),
		KBD_LAIR(12411, 2271, 4681);




		private final int tab;
		private final int x;
		private final int y;

		public int getTabId() {
			return tab;
		}

		public int getX() {
			return x;
		}

		public int getY() {
			return y;
		}

		TabType(int tab, int x, int y) {
			this.tab = tab;
			this.x = x;
			this.y = y;
		}
		}

	/**
	 * Operates the teleport tab
	 * 
	 * @param player
	 * @param item
	 */
	public static void operate(final Player player, int item) {
		for (TabType type : TabType.values()) {
			if (type.getTabId() == item) {
				if (System.currentTimeMillis() - player.lastTeleport < 3500)
					return;	
				if (!player.getPA().canTeleport("modern")) {
					return;
				}
				if (Boundary.isIn(player, Boundary.OUTLAST_HUT)) {
					player.sendMessage("Please leave the outlast hut area to teleport.");
					return ;
				}
				if (Boundary.isIn(player, Boundary.RAIDS_LOBBY) || Boundary.isIn(player, Boundary.RAIDS)) {
					player.sendMessage("Please leave the raids to teleport.");
					return ;
				}
				player.teleporting = true;
				player.getItems().deleteItem(type.getTabId(), 1);
				player.lastTeleport = System.currentTimeMillis();
				player.startAnimation(4731);
				player.gfx0(678);
				final int x = type.getX();
				final int y = type.getY();
				CycleEventHandler.getSingleton().addEvent(player, new CycleEvent() {

					@Override
					public void execute(CycleEventContainer container) {
						player.setTeleportToX(x);
						player.setTeleportToY(y);
						player.heightLevel = 0;
						player.teleporting = false;
						player.gfx0(-1);
						player.startAnimation(65535);
						container.stop();
					}

				}, 3);
			}
		}
	}

}

File: data/scripts/item/teleport/teleport_tablets.lua
require("ScriptLoader")

local Boundary = luajava.bindClass("io.kyros.model.entity.player.Boundary")


local TabType = {
    HOME = {tab = 8013, x = 3093, y = 3483},
    VARROCK = {tab = 8007, x = 3210, y = 3424},
    LUMBRIDGE = {tab = 8008, x = 3222, y = 3218},
    FALADOR = {tab = 8009, x = 2964, y = 3378},
    CAMELOT = {tab = 8010, x = 2757, y = 3477},
    ARDOUGNE = {tab = 8011, x = 2662, y = 3305},
    RIMMINGTON = {tab = 11741, x = 2956, y = 3217},
    TAVERLEY = {tab = 11742, x = 2896, y = 3456},
    POLLNIVENEACH = {tab = 11743, x = 3351, y = 2960},
    RELLEKKA = {tab = 11744, x = 2643, y = 3676},
    BRIMHAVEN = {tab = 11745, x = 2794, y = 3178},
    YANILLE = {tab = 11746, x = 2606, y = 3098},
    TROLLHEIM = {tab = 11747, x = 2888, y = 3676},
    ARCEUUS_LIBRARY = {tab = 19613, x = 1646, y = 3806},
    DRAYNOR_MANOR = {tab = 19615, x = 3109, y = 3346},
    SALVE_GRAVEYARD = {tab = 19619, x = 3434, y = 3461},
    FENKENSTRAIN_CASTLE = {tab = 19621, x = 3548, y = 3529},
    WEST_ARDOUGNE = {tab = 19623, x = 2538, y = 3305},
    HARMONY_ISLAND = {tab = 19625, x = 3794, y = 2851},
    CEMETERY = {tab = 19627, x = 2976, y = 3751},
    BARROWS = {tab = 19629, x = 3565, y = 3306},
    APE_ATOLL = {tab = 19631, x = 2757, y = 2781},
    PADDEWWA = {tab = 12781, x = 3098, y = 9884},
    SENNTISTEN = {tab = 12782, x = 3322, y = 3336},
    KHARYRLL = {tab = 12779, x = 3492, y = 3471},
    LASSAR = {tab = 12780, x = 3006, y = 3471},
    DAREEYAK = {tab = 12777, x = 2966, y = 3695},
    CARRALLANGAR = {tab = 12776, x = 3156, y = 3666},
    ANNAKARL = {tab = 12775, x = 3288, y = 3886},
    GHORROCK = {tab = 12778, x = 2977, y = 3873},
    MOONCLAN = {tab = 24949, x = 2113, y = 3915},
    OURANIA = {tab = 24951, x = 3015, y = 5622},
    WATERBIRTH_ISLAND = {tab = 24953, x = 2527, y = 3740},
    BARBARIAN = {tab = 24955, x = 2519, y = 3571},
    KHAZARD = {tab = 24957, x = 2660, y = 3158},
    FISHING_GUILD = {tab = 24959, x = 2635, y = 3425},
    CATHERBY = {tab = 24961, x = 2804, y = 3433},
    ICE_PLATEAU = {tab = 24963, x = 2916, y = 3921},
    WILDY_RESOURCE = {tab = 12409, x = 3184, y = 3945},
    PIRATE_HUT = {tab = 12407, x = 3045, y = 3956},
    MAGE_BANK = {tab = 12410, x = 2538, y = 4716},
    CALLISTO = {tab = 12408, x = 3293, y = 3847},
    KBD_LAIR = {tab = 12411, x = 2271, y = 4681}
}

local function canTeleport(player)
    if (os.time() * 1000 - player.lastTeleport < 3500) then
        return false
    end
    if not player:getPA():canTeleport("modern") then
        return false
    end
    if Boundary:isIn(player, Boundary.OUTLAST_HUT) then
        player:sendGameMessage("Please leave the outlast hut area to teleport.")
        return false
    end
    if Boundary:isIn(player, Boundary.RAIDS_LOBBY) or Boundary:isIn(player, Boundary.RAIDS) then
        player:sendGameMessage("Please leave the raids to teleport.")
        return false
    end
    return true
end

function onItemAction(event)
    if not event:getOption() == 2 then
        return
    end
    local player = event:getPlayer()
    local itemId = event:getItem()

    for _, tabType in pairs(TabType) do
        if tabType.tab == itemId then
            if not canTeleport(player) then
                return
            end

            player.teleporting = true
            player:getItems():deleteItem(tabType.tab, 1)
            player.lastTeleport = os.time() * 1000
            player:startAnimation(4731)
            player:gfx0(678)
            local x = tabType.x
            local y = tabType.y

            TaskScheduler:scheduleTask(1, function()
                player:setTeleportToX(x)
                player:setTeleportToY(y)
                player.heightLevel = 0
                player.teleporting = false
                player:gfx0(-1)
                player:startAnimation(65535)
                player:sendGameMessage("You have been teleported to " .. tabType.x .. ", " .. tabType.y .. ".")
            end, false, GAME_TICK)

            return
        end
    end
end

ScriptLoader.autoRegister()
